pkg/backupdriver/backup_driver_controller_base.go

```diff
@@ -416,7 +424,73 @@ func (ctrl *backupDriverController) pvWorker() {
 func (ctrl *backupDriverController) svcSnapshotWorker() {
 }
 
+// cloneFromSnapshotWorker is the main worker for restore request.
 func (ctrl *backupDriverController) cloneFromSnapshotWorker() {
+	ctrl.logger.Infof("cloneFromSnapshotWorker: Enter cloneFromSnapshotWorker")
+
+	key, quit := ctrl.cloneFromSnapshotQueue.Get()
+	if quit {
+		return
+	}
+	defer ctrl.cloneFromSnapshotQueue.Done(key)
+
+	if err := ctrl.syncCloneFromSnapshotByKey(key.(string)); err != nil {
+		// Put cloneFromSnapshot back to the queue so that we can retry later.
+		ctrl.cloneFromSnapshotQueue.AddRateLimited(key)
+	} else {
+		ctrl.cloneFromSnapshotQueue.Forget(key)
+	}
+}
+
+// syncCloneFromSnapshotByKey processes one CloneFromSnapshot CRD
+func (ctrl *backupDriverController) syncCloneFromSnapshotByKey(key string) error {
+	ctrl.logger.Infof("syncCloneFromSnapshotByKey: Started CloneFromSnapshot processing %s", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		ctrl.logger.Errorf("Split meta namespace key of CloneFromSnapshot %s failed: %v", key, err)
+		return err
+	}
+
+	cloneFromSnapshot, err := ctrl.cloneFromSnapshotLister.CloneFromSnapshots(namespace).Get(name)
```

Comment:
Ideally, lister should work just fine if the controller only respond to `AddFunc`. 

However, we did see some duplicate SnapshotManager.CreateSnapshot API calls before the fix, https://github.com/vmware-tanzu/velero-plugin-for-vsphere/pull/130/files. Please verify the commit multiple times and be careful about the case of duplicated calls.

Additionally, `AddFunc` will be triggered again if BackupDriver Pod crash and BackupDriver restart again. If we don't filter out terminal states, "Snapshotted" Snapshot CR or "Completed" CloneFromSnapshot CR will be handled once again.

Reply:
Added check to only create new if phase is New or Retry for clone.